karapace/protobuf/proto_normalizations.py

```diff
@@ -0,0 +1,220 @@
+"""
+Copyright (c) 2024 Aiven Ltd
+See LICENSE for details
+"""
+from karapace.protobuf.enum_constant_element import EnumConstantElement
+from karapace.protobuf.enum_element import EnumElement
+from karapace.protobuf.extend_element import ExtendElement
+from karapace.protobuf.field_element import FieldElement
+from karapace.protobuf.group_element import GroupElement
+from karapace.protobuf.message_element import MessageElement
+from karapace.protobuf.one_of_element import OneOfElement
+from karapace.protobuf.option_element import OptionElement
+from karapace.protobuf.proto_file_element import ProtoFileElement
+from karapace.protobuf.rpc_element import RpcElement
+from karapace.protobuf.service_element import ServiceElement
+from karapace.protobuf.type_element import TypeElement
+from karapace.typing import StrEnum
+from typing import List
+
+
+class ProtobufNormalisationOptions(StrEnum):
+    sort_options = "sort_options"
+
+
+def sort_by_name(element: OptionElement) -> str:
+    return element.name
+
+
+def type_field_element_with_sorted_options(type_field: FieldElement) -> FieldElement:
+    sorted_options = None if type_field.options is None else list(sorted(type_field.options, key=sort_by_name))
+    return FieldElement(
+        location=type_field.location,
+        label=type_field.label,
+        element_type=type_field.element_type,
+        name=type_field.name,
+        default_value=type_field.default_value,
+        json_name=type_field.json_name,
+        tag=type_field.tag,
+        documentation=type_field.documentation,
+        options=sorted_options,
+    )
+
+
+def enum_constant_element_with_sorted_options(enum_constant: EnumConstantElement) -> EnumConstantElement:
+    sorted_options = None if enum_constant.options is None else list(sorted(enum_constant.options, key=sort_by_name))
+    return EnumConstantElement(
+        location=enum_constant.location,
+        name=enum_constant.name,
+        tag=enum_constant.tag,
+        documentation=enum_constant.documentation,
+        options=sorted_options,
+    )
+
+
+def enum_element_with_sorted_options(enum_element: EnumElement) -> EnumElement:
```

Comment:
It would be valuable if we can encode normalization in the type system, so that instead of having functions like `(EnumElement) -> EnumElement` here we would have functions like `(EnumElement) -> NormalizedEnumElement`. That way we can potentially make use of the type system to bring confidence we always have normalized items when normalization is required. Presumably we would also need an additional type at the "top".

Reply:
The entrypoint to normalize is `def normalize_options_ordered`

How about having
```python
class NormalizedProtobufSchema(ProtobufSchema): ...

class NormalizedProtoFileElement(ProtoFileElement):
     
class NormalizedEnumElement(EnumElement): ...

...

Usage would be at the `schema_models.parse_protobuf_schema_definition`:
```python
protobuf_schema = ProtobufSchema(schema_definition, references, dependencies) if not normalize else NormalizedProtobufSchema(schema_definition, references, dependencies)
    if validate_references:
    ...
```


This would give the same interface for normalized versions and give typing to some extent.

Reply:
When loading already registered schemas from `_schemas` how do we know if those are normalized or not, and what is gained by having normalized schema type? I don't see much value here.

For example if loading earlier registered schema from `_schemas` we do not know if it was registered as with normalize flag or not. What is the type of it, and if compared to newly registering schema with normalize flag true or false, how does these types affect equality check?

Reply:
You are right. The property of normalization is not essential nor does provide any value at consuming side. Considering your comment and I also think that this is functional property at registration.

Reply:
I don't think so, they would be just different and not equal?